department-of-veterans-affairs/caseflow
# Adding an async loader component

![image](https://user-images.githubusercontent.com/25331532/33785842-df13f342-dc34-11e7-98d6-9b1fdec25136.png)

This is one of those "not sure this is a real improvement" PRs. :smile:

### Benefits
1. Code in Reader and Certification that tracks request failure state is removed.
1. Reader gets "loading is taking longer than expected" functionality for free.
1. Reader gets request timeout handling for free, which it previously did not have. 
1. Certification polling code is cleaned up considerably (IMO).

### Cons
1. Loader timeout can conflict with `ApiUtil` timeout.
1. This may be adding more indirection that's not really worth it.

## To Test
See that all Reader and Certification loading functionality continues to work under the following scenarios:

1. Data loads successfully and quickly.
1. Data loads successfully and slowly.
1. Data fails to load quickly.
1. Data fails to load slowly.

## TODO
Add tests.

## Notes

This is a bit over the line limit, but some large chunks are just cut/pastes of identical code. If you take those out, I'm pretty sure this is a net decrease in LOC. And if you view with `?w=1`, the LOC is greatly reduced.

I left some comments to help review that got swallowed by GH, so be sure to check them out if you'd like.

Connects #2794.
